torch/csrc/jit/runtime/register_prim_ops.cpp

```diff
@@ -627,6 +627,10 @@ RegisterOperators reg(
          TORCH_SELECTIVE_SCHEMA("aten::len.t(t[] a) -> int"),
          listLen,
          aliasAnalysisFromSchema()),
+     OperatorGenerator(
```

Comment:
Speaking of optimizable IR, whenever we add a new operator there will not be optimizations that know how to reason about it. Right now if you write:
```
def foo(x):
    l = [x]
    if len(x) == 1:
        return 0
    else:
        return 1
```
It will get optimized away because we have an [optimization pass](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/peephole_list_idioms.cpp) that tries to optimize away list operators. It was added because it cleans up control flow and makes it easier for other optimizations: https://github.com/pytorch/pytorch/pull/37612. I also have concrete plans to use the fact that we can optimize lists well to for shape analysis. 

Right now, the semantically equivalent `if l` will not get optimized, because it will call `aten::Bool` instead of `aten:len` and `aten::eq`. For optimization purposes, it would be great to compose the operators for types we care about optimizing in terms of our existing ones.

Reply:
More generally, it's good to not expand the IR too much if we can avoid it, because it does things like keep the mobile binary smaller and would make the IR easier to compile to LLVM if we ever want to do that...

Reply:
Does it make more sense to canonicalize these ops as `len(c) > 0` or generate them as `len(c) > 0` to start with? I'm about to update the PR to do the latter.

Reply:
Up to you. You can either add special handling in ir_emitter.cpp in `emitToBool` or add a case in https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/normalize_ops.cpp so that they get decomposed. 

Ideally, we'd be able to register these ops in `builtin_functions.cpp`. Plug for the hack-a-month idea i have listed [here](https://fb.quip.com/7UcNAhDrHQls). But that's longer term, probably shouldn't block this PR.

## Define Generic Python Builtins Using TorchScript Syntax (Add Support for TypeVars)

We would like to support python builtins such as  max, any, or iterttools.product. in a generic way that works for all builtin types & custom user types. 

An example of this would be:
```
def max(List[T] li) -> T
    assert li.size() > 0
    max_elem = li[0]
    for i in range(1, len(li)):
        if li[i] > max_elem:
            max_elem = li[i]
    return max_elem
```

We already support TypeVars on builtin operators (e.g.  aten::select(t[] li, int index) -> t  but we do not yet have a way of exposing this in a compiled function. If we did it would allow us to easily & maintainably increase our python coverage. Initially this would only extend to the usage in compiler builtins, but could eventually dovetail to exposing it to users.  
Schema To Allow us To Define Generic Implementations of Python Builtins